core/src/main/java/com/linecorp/armeria/client/DefaultClientFactory.java

```diff
@@ -82,6 +99,14 @@ final class DefaultClientFactory implements ClientFactory {
                 }
             }));
         }
+        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
```

Comment:
How about adding the shutdown hook only when the first `runnable` is added, so we don't add the shutdown hook unnecessarily?